### PR TITLE
[SPARK-21779][PYTHON] Simpler DataFrame.sample API in Python

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR make `DataFrame.sample(...)` can omit `withReplacement` defaulting `False`, consistently with equivalent Scala / Java API.

In short, the following examples are allowed:

```python
>>> df = spark.range(10)
>>> df.sample(0.5).count()
7
>>> df.sample(fraction=0.5).count()
3
>>> df.sample(0.5, seed=42).count()
5
>>> df.sample(fraction=0.5, seed=42).count()
5
```

In addition, this PR also adds some type checking logics as below:

```python
>>> df = spark.range(10)
>>> df.sample().count()
...
TypeError: withReplacement (optional), fraction (required) and seed (optional) should be a bool, float and number; however, got [].
>>> df.sample(True).count()
...
TypeError: withReplacement (optional), fraction (required) and seed (optional) should be a bool, float and number; however, got [<type 'bool'>].
>>> df.sample(42).count()
...
TypeError: withReplacement (optional), fraction (required) and seed (optional) should be a bool, float and number; however, got [<type 'int'>].
>>> df.sample(fraction=False, seed="a").count()
...
TypeError: withReplacement (optional), fraction (required) and seed (optional) should be a bool, float and number; however, got [<type 'bool'>, <type 'str'>].
>>> df.sample(seed=[1]).count()
...
TypeError: withReplacement (optional), fraction (required) and seed (optional) should be a bool, float and number; however, got [<type 'list'>].
>>> df.sample(withReplacement="a", fraction=0.5, seed=1)
...
TypeError: withReplacement (optional), fraction (required) and seed (optional) should be a bool, float and number; however, got [<type 'str'>, <type 'float'>, <type 'int'>].
```

## How was this patch tested?

Manually tested, unit tests added in doc tests and manually checked the built documentation for Python.

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -659,19 +659,77 @@ def distinct(self):
...
+        :param fraction: Fraction of rows to generate, range [0.0, 1.0].
+        :param seed: Seed for sampling (default a random seed).
+
         .. note:: This is not guaranteed to provide exactly the fraction specified of the total
             count of the given :class:`DataFrame`.
 
-        >>> df.sample(False, 0.5, 42).count()
-        2
-        """
-        assert fraction >= 0.0, "Negative fraction value: %s" % fraction
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: I removed this as it looks checked in Scala / Java side:

```python
>>> df.sample(fraction=-0.1).count()
...
pyspark.sql.utils.IllegalArgumentException: u'requirement failed: Sampling fraction (-0.1) must be on interval [0, 1] without replacement'
```

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -659,19 +659,77 @@ def distinct(self):
...
-        seed = seed if seed is not None else random.randint(0, sys.maxsize)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: I also removed `random.randint(0, sys.maxsize)` and tried to directly call Scala / Java side one.

**Comment by HyukjinKwon** [comment]
cc @rxin. Dose this make sense to you?

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -659,19 +659,77 @@ def distinct(self):
...
+
+        if not (is_withReplacement_set
+                or is_withReplacement_omitted_kwargs
+                or is_withReplacement_omitted_args):
+            argtypes = [
+                str(type(arg)) for arg in [withReplacement, fraction, seed] if arg is not None]
+            raise TypeError(
+                "withReplacement (optional), fraction (required) and seed (optional)"
+                " should be a bool, float and number; however, "
+                "got %s." % ", ".join(argtypes))
```
Comment (viirya) [inline python/pyspark/sql/dataframe.py]: By this change, all three parameters can be `None` by default, `argtypes` seems to be an empty list here?

Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Yea, it looks so. Let me try to improve this message.

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: I'd do the check in python, so the error message is more clear.

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -659,19 +659,77 @@ def distinct(self):
...
+        >>> df.sample(False, fraction=1.0).count()
+        10
+        >>> df.sample("a").count()
+        Traceback (most recent call last):
+            ...
+        TypeError:...
+        >>> df.sample(seed="abc").count()
+        Traceback (most recent call last):
+            ...
+        TypeError:...
```
Comment (felixcheung) [inline python/pyspark/sql/dataframe.py]: maybe we don't do the error cases here in doctest, but move them to unit test instead?
also these cases aren't really that meaningfully different to me as an user....?
```
        >>> df.sample(0.5, 3).count()
 +        4
 +        >>> df.sample(fraction=0.5, seed=3).count()
 +        4
 +        >>> df.sample(1.0).count()
 +        10
 +        >>> df.sample(fraction=1.0).count()
 +        10
 +        >>> df.sample(False, fraction=1.0).count()
 +        10
```

Comment (rxin) [inline python/pyspark/sql/dataframe.py]: that makes sense! doc tests are examples users can follow

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Hm.. wouldn't we better avoid duplicating expression requirement? It looks I should do:

https://github.com/apache/spark/blob/5ad1796b9fd6bce31bbc1cdc2f607115d2dd0e7d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L714-L722

within Python side. I have been thinking of avoiding it if the error message makes sense to Python users (but not the case of exposing non-Pythonic error messages, for example, Java types `java.lang.Long` in the error message) although I understand it is better to throw an exception ahead before going to JVM.

**Comment by HyukjinKwon** [comment]
https://github.com/apache/spark/pull/18999#discussion_r134131441 looks hidden. I addressed the other comment.

**Comment by HyukjinKwon** [comment]
@rxin, would you maybe have some opinion on https://github.com/apache/spark/pull/18999#discussion_r134131441 (avoiding fraction checking within Python side)?

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: yea it'd be better to have python handle the simpler error checking.

**Comment by HyukjinKwon** [comment]
cc @holdenk and @ueshin, could you maybe take a look when you have some time?

**Comment by ueshin** [comment]
LGTM.

**Review by felixcheung — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Thank you @viirya, @felixcheung, @rxin and @ueshin.

**Pull request closed without merging**